bin/cheritest/cheritest_sentries.c

```diff
@@ -0,0 +1,86 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2020 SRI International
+ * All rights reserved.
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory (Department of Computer Science and
+ * Technology) under DARPA contract HR0011-18-C-0016 ("ECATS"), as part of the
+ * DARPA SSITH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+
+#if !__has_feature(capabilities)
+#error "This code requires a CHERI-aware compiler"
+#endif
+
+#include <sys/param.h>
+#include <sys/syscall.h>
+#include <sys/sysctl.h>
+
+#include <unistd.h>
+
+#include "cheritest.h"
+
+static void
+check_fptr(void *fptr)
+{
+	register_t perms;
+
+	perms = cheri_getperm(fptr);
+
+	CHERITEST_VERIFY(cheri_gettag(fptr));
+	/* Check that execute is present and store permissions aren't */
+	CHERITEST_VERIFY2((perms & CHERI_PERM_EXECUTE) == CHERI_PERM_EXECUTE,
+	    "perms %jx (execute missing)", perms);
+	CHERITEST_VERIFY2((perms & CHERI_PERM_STORE) == 0,
+	    "perms %jx (store present)", perms);
+	CHERITEST_VERIFY2((perms & CHERI_PERM_STORE_CAP) == 0,
+	    "perms %jx (storecap present)", perms);
+	CHERITEST_VERIFY2((perms & CHERI_PERM_STORE_LOCAL_CAP) == 0,
+	    "perms %jx (store_local_cap present)", perms);
+
+	CHERITEST_VERIFY2(cheri_gettype(fptr) == CHERI_OTYPE_SENTRY,
+	    "otype %jx (expected %jx)", cheri_gettype(fptr),
+	    (uintmax_t)CHERI_OTYPE_SENTRY);
+
+	cheritest_success();
+}
+
+void
+test_sentry_libc(const struct cheri_test *ctp __unused)
+{
+	unsigned int (*fptr)(unsigned int seconds) = sleep;
+
+	check_fptr((void *)fptr);
```

Comment:
I think I'd prefer these to be `uintptr_t` rather than `void *`; the former is more in the spirit of the C standard. But not a strong opinion, and others may feel differently.